akka-persistence/src/main/scala/akka/persistence/PersistentFSMActor.scala

```diff
@@ -0,0 +1,113 @@
+/**
+ * Copyright (C) 2009-2014 Typesafe Inc. <http://www.typesafe.com>
+ */
+
+package akka.persistence
+
+import akka.actor.ActorLogging
+import akka.persistence.temp.{FSM, AbstractFSM, AbstractLoggingFSM}
+
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * FSM actor implementation with persistent state
+ *
+ * Supports the usual [[akka.actor.FSM]] functionality with additional persistence features.
+ * State and State Data are persisted on every state change.
+ * FSM is identified by 'persistenceId' value.
+ * Persistence execution order is: persist -> wait for ack -> apply state. Incoming messages are deferred until the state is applied.
+ */
+trait PersistentFsmActor[S,D,E] extends PersistentActor with FSM[S,D,E] with ActorLogging {
+  import akka.persistence.PersistentFsmActor._
+
+  /**
+   * Override this handler to define the action on Domain Event during recovery
+   *
+   * @param domainEvent domain event to apply
+   * @param currentData state data of the previous state
+   * @return updated state data
+   */
+  def applyEvent(domainEvent: E, currentData: D): D
+
+  /**
+   * Override this handler to define the action on recovery completion
+   */
+  def onRecoveryCompleted(): Unit = {}
+
+  /**
+   * After recovery events are handled as in usual FSM actor
+   */
+  override def receiveCommand: Receive = {
+    super[FSM].receive
+  }
+
+  /**
+   * Discover the latest recorded state
+   */
+  override def receiveRecover: Receive = {
+    case StateChangeEvent(state: S, None, timeout)                  => startWith(state, stateData, timeout)
+    case StateChangeEvent(state: S, Some(domainEvent: E), timeout)  => startWith(state, applyEvent(domainEvent, stateData), timeout)
+    case RecoveryCompleted                                          =>
+      initialize()
+      onRecoveryCompleted()
+  }
+
+  /**
+   * Persist FSM State and FSM State Data
+   */
+  override private[akka] def applyState(nextState: State): Unit = {
+    persist(StateChangeEvent(nextState.stateName, nextState.domainEvent, nextState.timeout))(ev => {
+      nextState.domainEvent match {
+        case None           => super.applyState(nextState)
+        case Some(event: E) => super.applyState(nextState using applyEvent(event, stateData))
+      }
+      nextState.afterTransitionDo(stateData)
+    })
+  }
+}
+
+
+
+object PersistentFsmActor {
+  /**
+   * Base persistent event class
+   */
+  private sealed trait PersistentFsmEvent
+
+  /**
+   * Persisted on state change
+   *
+   * @param state FSM state
+   * @param domainEvent FSM domain event
+   */
+  private case class StateChangeEvent[S, E](state: S, domainEvent: Option[E], timeout: Option[FiniteDuration]) extends PersistentFsmEvent
```

Comment:
How will we serialize `S`?
Java serialization is to dangerous, since S can be anything.
It would be too inconvenient for the user to mandate custom serializer for each state.

One idea is to demand that `S` is a subclass of:

    trait FSMState {
        def identifier: String
    }

Then we can serialize only the identifier, and maintain a mapping between identifier and corresponding `S`.

WDYT?